Adjust interception of requests for specific shard IDs

AthenaEryma opened the pull request:
Some index requests target shard IDs specifically, which may not match the indices that the request targets as given by IndicesRequest, which requires a different interception strategy in order to make sure those requests are handled correctly in all cases.

This PR adds and interface allowing requests to report the shard IDs they target as well as the index names, and adjusts the interception of those requests as appropriate to handle those shard IDs in the cases where they are relevant.

AthenaEryma committed:
Messing around

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

AthenaEryma committed:
Do it right

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

AthenaEryma committed:
Bit of cleanup

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticsearchmachine commented:
Hi @gwbrown, I've created a changelog YAML for you.

AthenaEryma committed:
Update docs/changelog/101656.yaml

AthenaEryma committed:
More cleanup

AthenaEryma committed:
Merge remote-tracking branch 'gwbrown/rcs/ssr-auth' into rcs/ssr-auth

kpollich reviewed (approved):
Fleet codeowner 🚀

AthenaEryma committed:
Adjust interface name and remove default implementation

jakelandis committed:
example

jakelandis committed:
more

AthenaEryma committed:
Merge branch 'main' into rcs/ssr-auth

# Conflicts:
#	x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityFcActionAuthorizationIT.java

AthenaEryma committed:
Merge remote-tracking branch 'jake/jake_test1' into rcs/ssr-auth

AthenaEryma committed:
Hacking on the test draft

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

AthenaEryma committed:
post-merge fixup

AthenaEryma committed:
Test polishing + convert to manual allowlist

AthenaEryma committed:
Fill allowlist + more polishing

AthenaEryma committed:
Merge branch 'main' into rcs/ssr-auth

AthenaEryma committed:
Revert unnecessarily broad request changes

gwbrown commented:
@jakelandis I believe I've addressed the concerns we discussed, could you give me a review?

jakelandis commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java:
```diff
@@ -0,0 +1,121 @@
+
+        // These actions do not have any references to shard IDs in their requests.
+        ClusterSearchShardsAction.NAME,
+        TransportSearchShardsAction.NAME
```
this action is associated with ShardSearchRequest which does have shardIDs in their request.

jakelandis commented on test/framework/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequestHelper.java:
```diff
@@ -0,0 +1,20 @@
+
+    private SingleShardRequestHelper() {}
+
+    public static void setInternalShardId(SingleShardRequest<?> singleShardRequest, ShardId shardId) {
```
Seems a bit off to introduce production code to allow invalid states only used in tests. I have an alternative idea will ping you about.

jakelandis commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return Set.of(LocalStateCompositeXPackPlugin.class);
```
This results in only checking the (core) xpack plugin and (core) server code's registered actions. Do you think it is worth pull this out to it's own qa project and have that qa project depend on a large % of the core/modules and x-pack/plugins. I understand this is best effort only but pulling this out to it's own project will get coverage similar to the default distribution and help to identify any new actions (unless the new actions are part of a new plugin/module)

gwbrown replied:
Blah, whoops, for some reason I thought switching this to `LocalStateCompositeXPackPlugin` addressed that but upon looking again you are correct and it does not. I'll look into this and look to get this pulled out somewhere we get more coverage.

(I really dislike this part of our test infrastructure and find it quite confusing, which plugins are included/accessible where)

jakelandis reviewed (commented):
Looking good... I wish there were better ways to test, but what you have here is seems the best we can do with what we have. 

A couple comments.

AthenaEryma committed:
Make action check more complete

AthenaEryma committed:
Implement interface for downsample read requests

AthenaEryma committed:
Merge branch 'main' into rcs/ssr-auth

AthenaEryma committed:
Spotless

AthenaEryma committed:
Remove test-only production code

This commit switches the test to stream a malformed get request directly, rather than using a real get request so that we don't have to have test-only production code.

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into rcs/ssr-auth

gwbrown commented:
@jakelandis I yes-and'd idea into "create a writeable that streams the equivalent to a get request"; I think that makes it a bit more maintainable while still letting us remove the test-only production code. I've also expanded the test coverage to cover x-pack, which had a couple implications I Slack'd you about.

jakelandis reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.13  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 101656`
